wakuv2/filter_manager.go

```diff
@@ -0,0 +1,360 @@
+	getFilter        func(string) *common.Filter
+	onNewEnvelopes   func(env *protocol.Envelope) error
+	peers            []peer.ID
+	logger           *zap.Logger
+	settings         settings
+	node             *node.WakuNode
+}
+
+func newFilterManager(ctx context.Context, logger *zap.Logger, getFilterFn func(string) *common.Filter, settings settings, onNewEnvelopes func(env *protocol.Envelope) error, node *node.WakuNode) *FilterManager {
+	// This fn is being mocked in test
+	mgr := new(FilterManager)
+	mgr.ctx = ctx
+	mgr.logger = logger
+	mgr.getFilter = getFilterFn
+	mgr.onNewEnvelopes = onNewEnvelopes
+	mgr.filterSubs = make(FilterSubs)
+	mgr.eventChan = make(chan FilterEvent)
+	mgr.peers = make([]peer.ID, 0)
+	mgr.settings = settings
+	mgr.node = node
+	mgr.isFilterSubAlive = func(sub *subscription.SubscriptionDetails) error {
+		ctx, cancel := context.WithTimeout(ctx, pingTimeout)
+		defer cancel()
+		return mgr.node.FilterLightnode().IsSubscriptionAlive(ctx, sub)
+	}
+
+	return mgr
+}
+
+func (mgr *FilterManager) runFilterLoop(wg *sync.WaitGroup) {
+	defer wg.Done()
+	// Use it to ping filter peer(s) periodically
+	ticker := time.NewTicker(5 * time.Second)
+	defer ticker.Stop()
+
+	// Populate filter peers initially
+	mgr.peers = mgr.findFilterPeers() // ordered list of peers to select from
+
+	for {
+		select {
+		case <-mgr.ctx.Done():
+			return
+		case <-ticker.C:
+			mgr.peers = mgr.findFilterPeers()
+			mgr.pingPeers()
+		case ev := <-mgr.eventChan:
+			switch ev.eventType {
+
+			case FilterEventAdded:
+				mgr.filterSubs[ev.filterID] = make(subscription.SubscriptionSet)
+				mgr.resubscribe(ev.filterID)
+
+			case FilterEventRemoved:
+				for _, sub := range mgr.filterSubs[ev.filterID] {
+					if sub == nil {
+						// Skip temp subs
+						continue
+					}
+					go mgr.unsubscribeFromFilter(ev.filterID, sub)
+				}
+				delete(mgr.filterSubs, ev.filterID)
+
+			case FilterEventPingResult:
+				if ev.result {
+					break
+				}
+				if ev.filterID != "" {
+					// Full resubscribe, filter has no peers
+					mgr.logger.Debug("filter has no subs", zap.String("filterId", ev.filterID))
+					mgr.resubscribe(ev.filterID)
+					break
+				}
+				// Remove peer from list
+				for i, p := range mgr.peers {
+					if ev.peerID == p {
+						mgr.peers = append(mgr.peers[:i], mgr.peers[i+1:]...)
+						break
+					}
+				}
+				// Delete subs for removed peer
+				for filterID, subs := range mgr.filterSubs {
+					for _, sub := range subs {
+						if sub == nil {
+							// Skip temp subs
+							continue
+						}
+						if sub.PeerID == ev.peerID {
+							mgr.logger.Debug("FILTER sub is inactive", zap.String("filterId", filterID), zap.String("subID", sub.ID))
+							delete(subs, sub.ID)
+							go mgr.unsubscribeFromFilter(filterID, sub)
+						}
+					}
+					mgr.resubscribe(filterID)
+				}
+
+			case FilterEventSubscribeResult:
+				subs, found := mgr.filterSubs[ev.filterID]
+				if ev.result {
```

Comment:
`success` sounds better than result to me, a bit clearer

Reply:
Yes, ok.